### PR TITLE
Use a structure to keep the allocated ips pool.

### LINKED ISSUE #11624: Refactor ipallocator package to not use global state.
**Problem**
 
The package `daemon/networkdriver/ipallocator` uses global state for ip address allocation.
 
**Solution**
 
The `ipallocator` used by Docker to assign IP addresses to starting containers stores its state in the global scope (see the `allocatedIPs` var in `ipallocator.go`).
 
Refactor the package to avoid using global state, for example by providing a `func New() &IPAlocator` call where `IPAllocator` would hold the entire state.

### PR DESCRIPTION
**Pull request by calavera**
Fixes #11624.

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: daemon/networkdriver/ipallocator/allocator.go [hunk 1]
```diff
@@ -41,19 +41,24 @@ var (
 	ErrBadSubnet                = errors.New("network does not contain specified subnet")
 )
 
-var (
-	lock         = sync.Mutex{}
-	allocatedIPs = networkSet{}
-)
+type IPAllocator struct {
+	allocatedIPs networkSet
+	sync.Mutex
```
Comment (LK4D4) [inline daemon/networkdriver/ipallocator/allocator.go]: can we not expose `sync.Mutex`, because `Lock/Unlock` methods for ipallocator look weird.

Comment (calavera) [inline daemon/networkdriver/ipallocator/allocator.go]: definitely, I actually meant to not expose it but I forgot about it at the end :pensive:

**Comment by calavera** [comment]
@LK4D4 I've made the mutex private.

**Comment by LK4D4** [comment]
@calavera Your second commit is not signed. Can you squash it to first?

**Comment by calavera** [comment]
Rebased to have only one signed commit.

**Comment by icecrime** [comment]
Cool thanks :-)

LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**